### PR TITLE
Fix warning-errors and set Werror for Epetra in GCC 7.2.0 PR build

### PR DESCRIPTION
**Pull request by ZUUL42**
@trilinos/framework 
@trilinos/epetra 

## Description

Fixed warning-errors and set `-Werror` for Epetra in the GCC 7.2.0 PR build.
Wrapped variables Werror sees as unused that are only used by asserts in `#ifndef NDEBUG`.
Wrapped or changed some `int` as `unsigned int` to fix the `always false [-Werror=strict-overflow]` Werrors.

## Motivation and Context

This completes #4468 and is work toward Issue #3178, eventually setting `-Werror` globally.

## How Has This Been Tested?

This was tested sucessfully running the GCC 7.2.0 PR Jenkins job.

## Checklist

- [x] My commit messages mention the appropriate GitHub issue numbers.
- [x] My code follows the code style of the affected package(s).
- [ ] My change requires a change to the documentation.
- [x] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [x] All new and existing tests passed.
- [x] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

### CODE REVIEW

File: packages/epetra/src/Epetra_Util.h [hunk 1]
```diff
@@ -348,7 +348,7 @@ int Epetra_Util_insert_empty_positions(T*& array, int& usedLength,
     return(-1);
   }
 
-  if ((usedLength+numPositions) < allocatedLength) {
+  if (((unsigned int)usedLength+(unsigned int)numPositions) < (unsigned int)allocatedLength) {
```
Comment (mhoemmen) [inline packages/epetra/src/Epetra_Util.h]: I can guess why compilers warned on this, but I'm not a huge fan of the change.  It turns undefined behavior on overflow, into well defined but totally wrong behavior on overflow.  I'm not sure what change would be better, though.

File: packages/epetra/test/VbrMatrix/cxx_main.cpp [hunk 2]
```diff
@@ -1740,7 +1740,7 @@ int checkMergeRedundantEntries(Epetra_Comm& comm, bool verbose)
   int localProc = comm.MyPID();
 
   int myFirstRow = localProc*3;
-  int myLastRow = myFirstRow+2;
+  unsigned int myLastRow = myFirstRow+2;
```
Comment (mhoemmen) [inline packages/epetra/test/VbrMatrix/cxx_main.cpp]: Wait, woah, why is this `unsigned` here?

File: packages/epetra/test/VbrMatrix/cxx_main.cpp [hunk 3]
```diff
@@ -1781,7 +1781,7 @@ int checkMergeRedundantEntries(Epetra_Comm& comm, bool verbose)
   //FillComplete, which internally calls MergeRedundantEntries, the
   //matrix should contain 1.0 in each entry.
 
-  for(int i=myFirstRow; i<=myLastRow; ++i) {
+  for(unsigned int i=myFirstRow; i<=myLastRow; ++i) {
```
Comment (mhoemmen) [inline packages/epetra/test/VbrMatrix/cxx_main.cpp]: `unsigned` shouldn't be the right type for a global index in Epetra.

File: packages/epetra/test/VbrMatrix/cxx_main.cpp (hunk 2 above)
Comment (ZUUL42) [reply packages/epetra/test/VbrMatrix/cxx_main.cpp]: This took care of the Werror coming from lines 1758 & 1897:
`error: assuming signed overflow does not occur when assuming that (X + c) < X is always false [-Werror=strict-overflow]`

File: packages/epetra/src/Epetra_Util.h (hunk 1 above)
Comment (ZUUL42) [reply packages/epetra/src/Epetra_Util.h]: I'm very open to alternatives.

**Comment by ZUUL42** [comment]
All of the instances of `unsigned` in this PR are for the same error type.
`error: assuming signed overflow does not occur when assuming that (X + c) < X is always false [-Werror=strict-overflow]`
It seemed the least intrusive way I found around those.
See [CDash report](https://testing.sandia.gov/cdash/viewBuildError.php?buildid=4976007)
If you know of a better fix, I am all ears.
I was just hoping this would be good or at least good enough for now and can be modified in the future to something more appropriate.

**Comment by mhoemmen** [comment]
The proper type of global indices in Epetra is `int` (or `long long` with Epetra64).  It's a bit risky to mix signed and unsigned types with global indices.  Bugs tend to show up only at large scales.

The "`X + c < X` is always false" warning is interesting -- I see three distinct variables in that expression in the code.  Perhaps the compiler has deduced that two of the variables are the same.

File: packages/epetra/test/VbrMatrix/cxx_main.cpp (hunk 3 above)
Comment (ZUUL42) [reply packages/epetra/test/VbrMatrix/cxx_main.cpp]: The `unsigned int i` felt rather wrong, but seemed to do the trick. I suppose I was _hoping_ large scale edge cases were tested for.
I went back and found if I changed `int myLastRow` and `int col` from the `unsigned int` I set to `long long`. Then I could return to `int i`. That seems better.
I'm still looking for an alternative to the `if ((usedLength+numPositions) < allocatedLength)` `strict-overflow` one though.

Comment (mhoemmen) [reply packages/epetra/test/VbrMatrix/cxx_main.cpp]: > I suppose I was _hoping_ large scale edge cases were tested for.

That's ... optimistic.

Comment (mhoemmen) [reply packages/epetra/test/VbrMatrix/cxx_main.cpp]: > I went back and found if I changed `int myLastRow` and `int col` from the `unsigned int` I set to `long long`. Then I could return to `int i`. That seems better.

Local row indices have type `int` in Epetra.

Comment (mhoemmen) [reply packages/epetra/test/VbrMatrix/cxx_main.cpp]: > I'm still looking for an alternative to the `if ((usedLength+numPositions) < allocatedLength)` strict-overflow one though.

That's a tricky one.  I wonder if the compiler has figured out that two of those variables are actually the same thing.
